File: test_pow/test_coreml.m
#import <Foundation/Foundation.h>
#import <CoreML/CoreML.h>
#import <os/log.h>
#import <sodium.h>

// Debug flag
static BOOL debugMode = NO;

#define INPUT_SIZE 32         // 32 bytes per nonce input
#define VECTOR_SIZE 256       // 256 bits per sample (32 bytes * 8)
#define HIDDEN_SIZE 1024
#define BATCH_SIZE 2048
#define ROUNDS 64

// Compute binary vector from a 32-byte input to a 256-element float vector.
void compute_binary_vector(const uint8_t *input, float *binary_out) {
    for (int i = 0; i < VECTOR_SIZE; i++) {
        binary_out[i] = (float)((input[i / 8] >> (7 - (i % 8))) & 1);
    }
}

void generate_sequential_nonces(uint8_t *outputs, uint64_t start_nonce, int batch_size) {
    for (int i = 0; i < batch_size; i++) {
        // Zero out the full input (32 bytes per sample)
        memset(outputs + (i * INPUT_SIZE), 0, INPUT_SIZE);
        // Write sequential nonce to the last 8 bytes (LSB)
        uint64_t nonce = start_nonce + i;
        for (int j = 0; j < 8; j++) {
            outputs[(i * INPUT_SIZE) + (24 + j)] = (nonce >> (8 * (7 - j))) & 0xFF;
        }
    }
}

// Helper function to print vectors (assumes contiguous array of given size)
void print_vector(const char *label, float *vector, int size) {
    printf("%s: [", label);
    for (int i = 0; i < size; i++) {
        printf("%.0f", vector[i]);
        if (i < size - 1) printf(", ");
        if (i > 20) {
            printf("...");
            break;
        }
    }
    printf("]\n");
}

// For an MLMultiArray with shape [256, BATCH_SIZE] stored in C-order,
// the element for sample 'col' and bit 'row' is at offset: row * BATCH_SIZE + col.
int count_leading_zeros(MLMultiArray *output, NSInteger sampleIndex) {
    int zeros = 0;
    for (NSInteger j = 0; j < VECTOR_SIZE; j++) {
        NSInteger offset = j * BATCH_SIZE + sampleIndex;
        if ([output[offset] intValue] == 0) {
            zeros++;
        } else {
            break;
        }
    }
    return zeros;
}

@protocol MLFeatureProvider;

@interface InputFeatureProvider : NSObject <MLFeatureProvider>
@property (nonatomic, strong) MLMultiArray *input;
@property (nonatomic, strong) NSSet<NSString *> *featureNames;
@end

@implementation InputFeatureProvider
- (instancetype)initWithBatchSize:(NSInteger)batchSize startNonce:(uint64_t)startNonce {
    self = [super init];
    if (self) {
        NSError *error = nil;
        
        // Create input array with shape [256, batchSize] (each column is one sample)
        self.input = [[MLMultiArray alloc] initWithShape:@[@(VECTOR_SIZE), @(batchSize)]
                                                 dataType:MLMultiArrayDataTypeFloat32
                                                    error:&error];
        if (error) {
            NSLog(@"Error creating input array: %@", error);
            return nil;
        }
        
        // Allocate temporary memory for input batch (32 bytes per sample)
        uint8_t *input_batch = (uint8_t *)malloc(batchSize * INPUT_SIZE);
        if (!input_batch) {
            NSLog(@"Memory allocation error");
            return nil;
        }
        generate_sequential_nonces(input_batch, startNonce, (int)batchSize);
        
        // Get pointer to MLMultiArray data
        float *binary_data = (float *)self.input.dataPointer;
        
        // For each sample, compute the binary vector and store it into the appropriate column.
        for (int b = 0; b < batchSize; b++) {
            uint8_t *sample_input = input_batch + (b * INPUT_SIZE);
            float sample_binary[VECTOR_SIZE] = {0};
            compute_binary_vector(sample_input, sample_binary);
            // Write the computed 256-bit vector into column b.
            for (int i = 0; i < VECTOR_SIZE; i++) {
                binary_data[i * batchSize + b] = sample_binary[i];
            }
        }
        
        // Debug output for the first sample (column 0)
        if (debugMode) {
            printf("Nonce: ");
            for (int i = 0; i < INPUT_SIZE; i++) {
                printf("%02x", input_batch[i]);
            }
            printf("\n");
            
            float first_input[VECTOR_SIZE];
            for (int i = 0; i < VECTOR_SIZE; i++) {
                first_input[i] = binary_data[i * batchSize + 0];
            }
            print_vector("First input vector", first_input, VECTOR_SIZE);
        }
        
        free(input_batch);
        
        self.featureNames = [NSSet setWithArray:@[@"input"]];
    }
    return self;
}

- (nullable MLFeatureValue *)featureValueForName:(NSString *)featureName {
    if ([featureName isEqualToString:@"input"]) {
        return [MLFeatureValue featureValueWithMultiArray:self.input];
    }
    return nil;
}
@end

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        // Parse arguments
        if (argc < 2) {
            NSLog(@"Usage: %s [-d] <difficulty>", argv[0]);
            return 1;
        }
        
        if (sodium_init() < 0) {
            NSLog(@"Error initializing libsodium");
            return 1;
        }
        
        // Parse arguments
        int argIndex = 1;
        if (strcmp(argv[argIndex], "-d") == 0) {
            debugMode = YES;
            argIndex++;
            if (argIndex >= argc) {
                NSLog(@"Missing difficulty parameter");
                return 1;
            }
        }
        
        int target_difficulty = atoi(argv[argIndex]);
        
        // Configure compute units
        MLModelConfiguration *config = [[MLModelConfiguration alloc] init];
        config.computeUnits = MLComputeUnitsAll;
        
        // Load and compile model
        NSString *modelPath = @"test_coreml.mlpackage";
        NSURL *modelURL = [NSURL fileURLWithPath:modelPath];
        NSError *error = nil;
        
        NSURL *compiledUrl = [MLModel compileModelAtURL:modelURL error:&error];
        if (error) {
            NSLog(@"Error compiling model: %@", error);
            return 1;
        }
        
        MLModel *model = [MLModel modelWithContentsOfURL:compiledUrl configuration:config error:&error];
        if (error) {
            NSLog(@"Error loading model: %@", error);
            return 1;
        }
        
        NSLog(@"Mining with difficulty: %d%@", target_difficulty, debugMode ? @" (Debug mode enabled)" : @"");
        
        NSInteger batchSize = BATCH_SIZE;
        __block uint64_t nonce = 0;
        __block uint64_t totalHashes = 0;
        __block NSDate *startTime = [NSDate date];
        __block int best_difficulty = 0;
        
        // Status display timer - 1 second intervals
        dispatch_source_t timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0,
                                                         dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0));
        dispatch_source_set_timer(timer, dispatch_time(DISPATCH_TIME_NOW, 0), NSEC_PER_SEC, 0);
        dispatch_source_set_event_handler(timer, ^{
            NSTimeInterval elapsed = -[startTime timeIntervalSinceNow];
            double hashrate = totalHashes / elapsed;
            double tops = (hashrate * (ROUNDS * HIDDEN_SIZE * HIDDEN_SIZE * 2 + HIDDEN_SIZE * VECTOR_SIZE * 4)) / 1e12;
            NSLog(@"Nonce: %llu | Hashrate: %.2f H/s | TOPS: %.2f | Best difficulty: %d",
                  nonce, hashrate, tops, best_difficulty);
        });
        dispatch_resume(timer);
        
        // Mining loop
        while (true) {
            @autoreleasepool {
                InputFeatureProvider *inputProvider = [[InputFeatureProvider alloc] initWithBatchSize:batchSize startNonce:nonce];
                if (!inputProvider) {
                    NSLog(@"Error creating input provider");
                    continue;
                }
                
                id<MLFeatureProvider> output = [model predictionFromFeatures:inputProvider error:&error];
                if (error) {
                    NSLog(@"Error during inference: %@", error);
                    continue;
                }
                
                // Get output feature (ensure the feature name matches your model)
                MLFeatureValue *outputFeature = [output featureValueForName:@"clip_65"];
                if (!outputFeature) {
                    NSLog(@"Could not find output feature");
                    continue;
                }
                
                MLMultiArray *outputArray = [outputFeature multiArrayValue];
                
                if (debugMode) {
                    float first_output[VECTOR_SIZE];
                    for (int j = 0; j < VECTOR_SIZE; j++) {
                        first_output[j] = [outputArray[j * batchSize + 0] floatValue];
                    }
                    print_vector("First output vector", first_output, VECTOR_SIZE);
                    
                    uint8_t first_output_bytes[32] = {0};
                    for (NSInteger j = 0; j < VECTOR_SIZE; j++) {
                        NSInteger bitIndex = 7 - (j % 8);
                        NSInteger byteIndex = j / 8;
                        first_output_bytes[byteIndex] |= (([outputArray[j * batchSize + 0] intValue] & 1) << bitIndex);
                    }
                    NSMutableString *outputHex = [NSMutableString string];
                    for (int j = 0; j < 32; j++) {
                        [outputHex appendFormat:@"%02x", first_output_bytes[j]];
                    }
                    NSLog(@"First output (hex): %@", outputHex);
                }
                
                // Check each sample (column) in the batch
                for (NSInteger i = 0; i < batchSize; i++) {
                    int zeros = count_leading_zeros(outputArray, i);
                    if (zeros > best_difficulty) {
                        best_difficulty = zeros;
                    }
                    
                    if (zeros >= target_difficulty) {
                        uint64_t solution_nonce = nonce + i;
                        NSLog(@"\nSolution found!");
                        NSLog(@"Nonce: %llu", solution_nonce);
                        NSLog(@"Leading zeros: %d", zeros);
                        
                        NSMutableString *inputHex = [NSMutableString string];
                        for (int j = 0; j < 24; j++) {
                            [inputHex appendString:@"00"];
                        }
                        for (int j = 0; j < 8; j++) {
                            [inputHex appendFormat:@"%02x", (uint8_t)((solution_nonce >> (8 * (7 - j))) & 0xFF)];
                        }
                        NSLog(@"Solution input (hex): %@", inputHex);
                        
                        NSMutableString *outputHex = [NSMutableString string];
                        uint8_t output_bytes[32] = {0};
                        for (NSInteger j = 0; j < VECTOR_SIZE; j++) {
                            NSInteger bitIndex = 7 - (j % 8);
                            NSInteger byteIndex = j / 8;
                            output_bytes[byteIndex] |= (([outputArray[j * batchSize + i] intValue] & 1) << bitIndex);
                        }
                        for (int j = 0; j < 32; j++) {
                            [outputHex appendFormat:@"%02x", output_bytes[j]];
                        }
                        NSLog(@"Model output (hex): %@", outputHex);
                        
                        dispatch_source_cancel(timer);
                        return 0;
                    }
                }
                
                nonce += batchSize;
                totalHashes += batchSize;
            }
        }
    }
    return 0;
}

File: test_pow/test_coreml.py
import argparse
import time
from typing import List, Tuple
import numpy as np
import coremltools as ct

# Constants matching the original implementation
INPUT_SIZE = 32        # 32 bytes per nonce input
VECTOR_SIZE = 256      # 256 bits per sample
HIDDEN_SIZE = 1024
BATCH_SIZE = 2048
ROUNDS = 64

class ProofOfWorkMiner:
    def __init__(self, model_path: str, target_difficulty: int, debug: bool = False):
        self.target_difficulty = target_difficulty
        self.debug = debug
        
        # Load CoreML model
        spec = ct.utils.load_spec(model_path)
        self.model = ct.models.MLModel(model_path, compute_units=ct.ComputeUnit.ALL)
        
        # Initialize tracking variables
        self.nonce = 0
        self.total_hashes = 0
        self.start_time = time.time()
        self.best_difficulty = 0
        
        # Pre-allocate arrays for batch processing
        self.binary_input = np.zeros((VECTOR_SIZE, BATCH_SIZE), dtype=np.float32)
        self.nonce_bytes = np.zeros((BATCH_SIZE, INPUT_SIZE), dtype=np.uint8)

    def prepare_batch(self) -> None:
        """Prepare a batch of inputs for the model."""
        # Generate sequential nonces efficiently using numpy
        nonces = np.arange(self.nonce, self.nonce + BATCH_SIZE, dtype=np.uint64)
        
        # Convert nonces to bytes (last 8 bytes of each 32-byte input)
        for i in range(8):
            self.nonce_bytes[:, 24 + i] = (nonces >> (8 * (7 - i))) & 0xFF
            
        # Convert to binary representation efficiently
        bits = np.unpackbits(self.nonce_bytes, axis=1)
        self.binary_input = bits.reshape(BATCH_SIZE, -1)[:, :VECTOR_SIZE].T.astype(np.float32)

    def run_inference(self) -> np.ndarray:
        """Run model inference on a batch of inputs."""
        # Create input dictionary and run inference
        outputs = self.model.predict({'input': self.binary_input})
        
        # Get output array and ensure correct shape
        output_array = outputs['clip_65']
        return np.asarray(output_array).T

    def count_leading_zeros(self, output: np.ndarray) -> int:
        """Count leading zeros in binary output."""
        nonzero_indices = np.nonzero(output)[0]
        return nonzero_indices[0] if len(nonzero_indices) > 0 else VECTOR_SIZE

    def print_debug_info(self, binary_output: np.ndarray):
        """Print debug information for the first sample in batch."""
        print(f"Nonce: {bytes(self.nonce_bytes[0]).hex()}")
        print(f"Input vector: [{', '.join(map(str, self.binary_input[:, 0][:20]))}...]")
        print(f"Output vector: [{', '.join(map(str, binary_output[0, :20]))}...]")
        
        # Convert output bits to bytes
        output_bytes = np.packbits(binary_output[0].astype(bool))
        print(f"Output (hex): {bytes(output_bytes).hex()}")

    def print_status(self):
        """Print current mining status."""
        elapsed = time.time() - self.start_time
        hashrate = self.total_hashes / elapsed
        # Calculate TOPS (Tera Operations Per Second)
        tops = (hashrate * (ROUNDS * HIDDEN_SIZE * HIDDEN_SIZE * 2 + HIDDEN_SIZE * VECTOR_SIZE * 4)) / 1e12
        print(f"Nonce: {self.nonce} | Hashrate: {hashrate:.2f} H/s | "
              f"TOPS: {tops:.2f} | Best difficulty: {self.best_difficulty}")

    def check_solution(self, binary_output: np.ndarray, batch_index: int) -> bool:
        """Check if a solution meets the target difficulty."""
        zeros = self.count_leading_zeros(binary_output[batch_index])
        self.best_difficulty = max(self.best_difficulty, zeros)
        
        if zeros >= self.target_difficulty:
            solution_nonce = self.nonce + batch_index
            print("\nSolution found!")
            print(f"Nonce: {solution_nonce}")
            print(f"Leading zeros: {zeros}")
            print(f"Solution input (hex): {bytes(self.nonce_bytes[batch_index]).hex()}")
            
            output_bytes = np.packbits(binary_output[batch_index].astype(bool))
            print(f"Model output (hex): {bytes(output_bytes).hex()}")
            return True
            
        return False

    def mine(self):
        """Main mining loop."""
        print(f"Mining with difficulty: {self.target_difficulty}"
              f"{' (Debug mode enabled)' if self.debug else ''}")
        
        last_status_time = time.time()
        
        while True:
            # Prepare batch using pre-allocated arrays
            self.prepare_batch()
            
            # Run inference
            binary_output = self.run_inference()
            
            # Debug output for first sample
            if self.debug:
                self.print_debug_info(binary_output)
            
            # Check each sample in batch
            for i in range(BATCH_SIZE):
                if self.check_solution(binary_output, i):
                    return
            
            # Update counters
            self.nonce += BATCH_SIZE
            self.total_hashes += BATCH_SIZE
            
            # Print status every second
            if time.time() - last_status_time >= 1.0:
                self.print_status()
                last_status_time = time.time()

def main():
    parser = argparse.ArgumentParser(description='CoreML Proof of Work Miner')
    parser.add_argument('difficulty', type=int, help='Target difficulty (number of leading zeros)')
    parser.add_argument('-d', '--debug', action='store_true', help='Enable debug output')
    parser.add_argument('--model', default='test_coreml.mlpackage', 
                      help='Path to CoreML model package')
    args = parser.parse_args()

    miner = ProofOfWorkMiner(args.model, args.difficulty, args.debug)
    miner.mine()

if __name__ == '__main__':
    main()
